test/test_core.h

```diff
@@ -28,16 +28,80 @@
 #include "picotest/picotest.h"
 #include "testhelper.h"
 #include "tiny_dnn/tiny_dnn.h"
-
+#if defined(USE_OPENCL) || defined(USE_CUDA)
+#include "third_party/CLCudaAPI/clpp11.h"
+#endif //defined(USE_OPENCL) || defined(USE_CUDA)
 using namespace tiny_dnn;
 
 namespace tiny_dnn {
+#if defined(USE_OPENCL) || defined(USE_CUDA)
+device_t device_type(size_t &platform, size_t &device) {
+    //check which platforms are available
+    auto platforms = CLCudaAPI::GetAllPlatforms();
+
+    // if no platforms - return -1
+    if (platforms.size() == 0)
+        return device_t::NONE;
+
+    for (auto p = platforms.begin(); p != platforms.end(); ++p)
+        for (size_t d = 0; d < p->NumDevices(); ++d) {
+            auto dev = CLCudaAPI::Device(*p, d);
+            if (dev.Type()=="GPU") {
+                platform = p - platforms.begin();
+                device = d;
+                return device_t::GPU;
+            }
+        }
+
+    for (auto p = platforms.begin(); p != platforms.end(); ++p)
+        for (size_t d = 0; d < p->NumDevices(); ++d) {
+            auto dev = CLCudaAPI::Device(*p, d);
+            if (dev.Type()=="CPU") {
+                platform = p - platforms.begin();
+                device = d;
+                return device_t::CPU;
+            }
+        }
+
```

Comment:
@edgarriba Well I was prototyping, and two loops was most straightforward approach which I considered readable enough.

It can be refactored like that:
```
	size_t cpu_platform, cpu_device;
	bool has_CPU=false;
    for (auto p = platforms.begin(); p != platforms.end(); ++p)
        for (size_t d = 0; d < p->NumDevices(); ++d) {
            auto dev = CLCudaAPI::Device(*p, d);
            if (dev.Type()=="GPU") {
                platform = p - platforms.begin();
                device = d;
                return device_t::GPU;
            }
            if (!has_CPU && dev.Type()=="CPU") {
                cpu_platform = p - platforms.begin();
                cpu_device = d;
                has_CPU = true;
            }
        }

	//no GPU found
    if (has_CPU) {
		platform = cpu_platform;
		device = cpu_device;
		return device_t::CPU;
	}
	
	//no CPUs or GPUs
	return device_t::NONE;
```

Do you think it's more readable and intuitive?